### ISSUE TITLE
How does squiggly access fields?

### ISSUE THREAD

**Issue by bsara** [bohnman/squiggly-filter-jackson#16]
How does Squiggly access fields on an object that is to be serialized to JSON, does it use reflection to get the field value directly, use the field's getter, or not handle this at all and just allow Jackson to handle the retrieval of values?

I ask because I was hoping to use this library in conjunction with Spring data JPA. However, I always get `null` when filtering on a lazy loaded nested object, if I don't specify any fields and include the lazy loaded nested object by default, then it is not `null`. I hoped that I could exclude these lazy loaded nested objects by default, and include them on a case-by-case basis. It would also ensure that the extra SQL queries only occur when I need them to.

**Comment by bohnman** [comment]
Squiggly does not access fields/getters on objects.  Squiggly is implemented as a Jackson PropertyFilter, specifically a SimpleBeanPropertyFilter, where Jackson calls SquigglyPropertyFilter.include with essentially the name of the property and the method returns true/false on whether to include that property.   The SquigglyFilter does not have access to the values of the object nor is it even aware that it is inside of a collection property on the parent object.  The only reflection that Squiggly performs itself is in BeanInfoIntrospector, but that is on the class, not on an instance.

I'm assuming that your JPA implementation is Hibernate.  Did you try using the [Jackson Hibernate Module](https://github.com/FasterXML/jackson-datatype-hibernate)?  It's been 7+ years since I've used JPA so I'm not sure if this would even solve your issue.

If you cannot solve the issue and you're up to providing a sample project that exhibits your described behavior using either an in-memory database or PostgreSQL, I'd be happy to help look into to finding a potential solution.

**Comment by bohnman** [comment]
@bsara - I have created an example that uses the Jackson Hibernate module to not include lazy collections by default unless you specify a Squiggly filter expression.  Please note that this required a slight code change to Squiggly, so you will need to upgrade your project to use version 1.3.7.

The example can be found [here](https://github.com/bohnman/squiggly-filter-jackson/tree/master/examples/spring-data-jpa-hibernate).

Let me know if this works for you.

**Comment by bsara** [comment]
@bohnman first of all, thank you so much for the quick response!

I actually ended up running through the squiggly source and learned for myself what you stated about how everything works. I realized that my problem was actually caused because I was using the Jackson Hibernate Module. I ended up being able to fix my issue if I enabled "force lazy loading" in the Hibernate module. We initially did not force lazy loading because things were being loaded that we didn't want loaded...but with Squiggly, now I'm applying the `@FullView` annotation to those nested lazy loaded fields, which keep the getters from being called and prevents lazy loading unless I specifically include the field in my query on the endpoint.

So, basically, Squiggly is awesome and has fixed yet another big issue for us. ...so, thanks!

I also really appreciate the sample application, I will take a look at it very soon.
